### ISSUE TITLE
Build: the fold filters a Held Gap once in `predict`, the value-level doors reduce to the Investable Mask, and the census tripwire

### ISSUE THREAD

**Issue by dcelisgarza** [dcelisgarza/PortfolioOptimisers.jl#856]
## Question

Build the first half of the decision of [#673](https://github.com/dcelisgarza/PortfolioOptimisers.jl/issues/673), ADR 0118: a fold filters a Held Gap once in `predict`, a value-level verb against a Prior Result reduces to the Investable Mask at its entry, a series the caller holds is documented and not checked, and one census test is the tripwire.

### The pieces

1. **The fold's filter.** One verb, `filter_held_gaps(w, X, strict)`, which runs over the investable columns alone (see piece 2), beside `calc_net_returns` in `src/17_NetReturnsDrawdowns.jl` or in the cross-validation base. It returns `X` itself when `X` is finite. Otherwise it returns a copy with `0` at every non-finite entry, and it names the Held Gaps, the pairs where `w[i]` is non-zero and `X[t, i]` is non-finite, through `strict_diagnostic`: a warning by default, an `ArgumentError` under `strict`. A zero weight at a gap is silent. No renormalisation. `attribution_net_returns` in `src/21_FactorAttributionRealised.jl:1250` is the pattern; consider making it call the new verb.
2. **`predict` views at the mask, then calls it once.** Both `StatsAPI.predict(res, rd, …)` methods in `src/20_Optimisation/02_CrossValidation/01_Base_CrossValidation.jl` (`:1391`, `:1427`) first view the test window, the weights and the fees at `res.imsk` (`nothing` views nothing), so the column of a non-investable asset is never read. That is the first decision of #674, ADR 0120, and ADR 0118 was rewritten to this order. They then filter the reduced window, and hand the filtered matrix to `calc_net_returns` **and** to `held_weights_result`, so the drift compounds on the same matrix. The held weights after the last observation expand back to the full length through `expand_investable_weights`, because the next fold's turnover reads them. A weight path over a `MatNum` follows the same rule per pair.
3. **`strict` travels as a field.** Every cross-validation estimator gains `strict::Bool = false` beside `wd` and `store_weight_path` (`KFold`, `CombinatorialCrossValidation`, `IndexWalkForward`, `DateWalkForward`, `MultipleRandomised`, and the search estimators forward it). `predict` and `fit_and_predict` take it as a keyword with the same default. `Pipeline` forwards it where it forwards `store_weight_path` (`src/23_Pipeline/03_Pipeline.jl:917`).
4. **The value-level doors reduce.** A third method family of `investable_reduction(pr, w, strict)` beside the ADR 0115 methods: `investable_mask(pr)` once; `nothing` returns its arguments; otherwise diagnose a held non-investable asset (`w[i] != 0` where the mask is `false`) through `strict_diagnostic`, and return the mask, `port_opt_view(pr, imsk)` and `view(w, imsk)`. Call it at the entry of `expected_risk(r, w::VecNum, pr)`, `expected_risk(rs::VecBaseRM, w, pr)`, the `MatNum` twins, `expected_return(ret, w, pr)`, `risk_contribution(r, w, pr)` and `factor_risk_contribution(r, w, pr)`. Every one gains `strict::Bool = false`. A per-asset answer expands through `expand_investable_weights(imsk, rc)`. A bare `X::MatNum` entry is unchanged.
5. **The funnel is unchanged.** `calc_net_returns` and `calc_net_asset_returns` stay the plain product. Their docstrings gain one paragraph: a non-finite entry poisons its observation, `0 * NaN` is `NaN`, and a gapped panel is scored through `predict(res, rd)`, which filters once.
6. **A series the caller holds is documented, not checked.** `expected_risk_from_returns`, `performance_summary(ret)`, `cumulative_returns` and `drawdowns` state that the series must be finite, that a tail measure on a gapped series answers a finite wrong number, and that `x[isfinite.(x)]` before the call reproduces the reference implementation's drop-per-column answer exactly, except its unbiased semi-variance. No `assert_all_finite` on these paths: the maintainer refused the cost.
7. **The glossary and the contract.** `CONTEXT.md` carries the Held Gap entry and the amended Precomputed-returns contract entry from the ADR commit. Keep the docstrings in the glossary's words.

### The tests

- **The tripwire, by reflection.** One new test file walks the concrete subtypes of `AbstractBaseRiskMeasure` (guard against probe types from a shared REPL, see the auto-memory). For each measure that has a value-level method, on a Prior Result over a synthetic panel whose fourth asset is non-investable:
  - with `w` zero on the dead asset, `expected_risk(r, w, pr)` equals `expected_risk(r, w[keep], pr_keep)` where `pr_keep` is the prior fitted on the panel with the dead column removed by hand, the oracle of `test/test_50_investable_reduction.jl`;
  - with `w` non-zero on the dead asset, one warning (`Test.collect_test_logs`) and the same answer as with that weight zeroed; `strict = true` throws `ArgumentError`;
  - `risk_contribution` returns the full length with exactly `0` at the dead asset.
- **The fold.** One `IndexWalkForward` over a synthetic panel with a non-investable asset on the training window and a delisting of an investable asset inside the test window: the dead asset's `NaN` column is never read and warns nothing, `pred.hw.w` has the full length with `0` at the dead asset, and for the delisting `pred.rd.X` is finite, exactly one warning names the asset, the series equals the hand-computed `sum_i w_i * (isfinite(X[t,i]) ? X[t,i] : 0) - fee`, an all-gap row gives `-fee`, and `strict = true` throws. A zero weight on the delisted asset warns nothing (`@test_logs min_level = Logging.Warn`).
- **The doors.** `expected_risk_from_returns` and `performance_summary` are not asserted on a gapped series. Pin the documented compaction identity instead: for one measure of each kernel class, `f(x[isfinite.(x)])` equals the reference implementation's number recorded in the test.

### The gates

Every new file pays the sweep: `julia --project=code_health code_health/sweep_check.jl --fetch`, and the `sweep-conform` skill before the commit. The complexity and size baselines of the touched files move. The ADR is written; amend it only if the build finds it wrong, and say so on this ticket.

Blocked by nothing. Blocks the plot ticket and the closing verification #677.

**Comment by dcelisgarza** [comment]
## Resolution

Built and landed on `dev` as commit `e968f2624f`.

A Held Gap is an `(observation, asset)` pair at which the weight is non-zero and the return is missing. It is what an asset that delists **inside** a test window makes. Two layers met it, and neither said so. Both now do.

### The fold

`StatsAPI.predict` takes its test window, its weights and its fees at the Investable Mask **first**, then zeroes the Held Gaps of the reduced window **once**, before the series is formed and before a Weight Drift compounds on it. That is ADR 0120's order, and it is what the two methods now do.

The pinned identity is

```
returns[t] == sum_i w_i * (isfinite(X[t, i]) ? X[t, i] : 0) - fee
```

The Held Weights record is computed on the reduced window and expanded on the way out, so `pred.hw.w` comes back on the caller's universe at full length. A zero-filled dead column at a zero weight makes the rebuilt drift path bit-identical, which is what let the record be computed reduced and expanded rather than computed twice.

`result_investable_mask` is how the fold reads the mask a result carries. It has a `nothing` fallback on `OptimisationResult`, a method on the JuMP family reading `res.pa.imsk`, and one on each of `NaiveOptimisationResult`, `StackingResult` and `SubsetResamplingResult` reading `res.imsk`.

### The value-level doors

`expected_risk`, `expected_return`, `risk_contribution` and `factor_risk_contribution` reduce to the Investable Mask at their entry and expand a per-asset answer back to the full length. The composites — `expected_ratio`, `expected_risk_ret_ratio`, `expected_sric`, `expected_risk_ret_sric` — reduce at the outermost entry, so a held non-investable asset is named **once** and not twice. `NoReturn` is left alone: it answers zero at any weight.

`factor_risk_contribution` is the one verb that also cuts the returns it regresses against, through `investable_returns_view`, because the loadings are fitted per asset.

### The pieces

  - `filter_held_gaps`, `held_gap_pairs` and `held_gap_msg` in `src/17_NetReturnsDrawdowns.jl`, beside `calc_net_returns`.
  - `investable_reduction` over a matrix, a Returns Result and a Prior Result, in the same file. `held_non_investable` and `investable_weights_view` in `src/13_Prior/01_Base_Prior.jl`.
  - `expand_investable_columns`, `expand_held_weights` and `expand_held_member` carry the fold's answer back to the caller's universe.
  - `strict` travels as a field of `KFold`, `CombinatorialCrossValidation`, `IndexWalkForward`, `DateWalkForward` and `MultipleRandomised`, and as a keyword through the search schemes and the pipeline, wherever `store_weight_path` already travels. `fold_evaluation` reads it.
  - `calc_net_returns` and `calc_net_asset_returns` stay the plain product. `expected_risk_from_returns`, `performance_summary(ret)`, `cumulative_returns` and `drawdowns` **document** the Precomputed-returns contract rather than checking it.
  - `CONTEXT.md` already carried the Held Gap and Precomputed-returns entries from the ADR commit, so it needed no change.

### The tests

`test/test_54_held_gap_filter.jl`, 309 passing. The oracle everywhere is the same problem with the non-investable asset removed **by hand**, which is the oracle `test/test_50_investable_reduction.jl` uses on the optimiser side.

Two tripwires, because a written list is exactly what a new type is forgotten in:

  - One walks every concrete `AbstractBaseRiskMeasure` and asserts the prior route reduces. It censuses 57 concrete types, of which 40 answer both sides of the oracle.
  - One censuses every concrete `OptimisationResult` that carries an `imsk` field and asserts it answers `result_investable_mask` rather than the fallback. **It fired during the rebase** and found the three families a sibling ticket had just given a mask to.

### Two deviations from the ticket

Both deliberate, and both are in the commit message.

  - `investable_reduction` takes four arguments, `(pr, w, fees, strict)`, not three. A per-asset `Fees` rides the same axis as the weights, so it must be cut by the same mask in the same call.
  - It lives in `src/17_NetReturnsDrawdowns.jl` rather than `src/13_Prior/`. `Fees` is defined in file 16, so a signature that names it cannot load in file 13.

### What was verified

  - `test_54_held_gap_filter` (309), `test_50_investable_reduction`, `test_24_cross_validation`, `test_11_fees_and_returns`, `test_19_factor_risk_contribution`, `test_21a_performance_summary`, `test_21b_factor_attribution`, `test_09_risk_measures`, `test_22c_views_fees`, `test_18r_value_level_multiplicity`, `test_29_risk_measure_compatibility`, `test_30_pipeline_base`, `test_33_pipeline_prediction_cv`, `test_34_pipeline_search_cv`, `test_37_train_test_split`, `test_14_naive_optimisation`, `test_42_combination_weight_stacking`, `test_21_subset_resampling_optimisarion`. All pass.
  - Every census gate: `test_26_docs`, `test_40`, `test_41`, `test_43`, `test_44`, `test_45`, `test_46`, `test_47`, `test_48`, `test_49`, `test_50_docs_sitemap`, `test_51`, `test_52`, `test_53`, `test_27`, `test_28`. All pass.
  - The doctests, run in a fresh process with the CI invocation. They pass.
  - Every `@docs` entry on the touched pages resolves to exactly one docstring, checked with `Documenter.DocSystem`. This matters because `investable_reduction` now carries two docstrings, so both its entries are qualified by signature.
  - The code-health gates: size, complexity, expansion and JET are all green.
  - Coverage of the added lines, measured with `--code-coverage` over the worktree `src/`: every added line that carries a counter is hit.

### The gates that moved

  - `sweep/manifest.toml`: four unit counts.
  - `code_health/size_baseline.toml`: four rises over the 500 ceiling.
  - `code_health/complexity_baseline.toml`: the `strict` argument on eight files.
  - `code_health/jet_baseline.toml`: a rise of two in `src/13_Prior/01_Base_Prior.jl` in the `plots_ext` run. Both reports are **false positives** inside existing inner constructors of `LowOrderPrior` and `HighOrderPrior`: an `@argcheck` establishes an invariant that inference cannot fold, and the next line reads a field of the value it pairs. The new seam merely made them reachable from the plotting extension. A Dismissal needs a new Rationale, which needs the maintainer, so route 3 of the gate's own instructions was taken. **#878** carries the reproduction, both stacks and the proposed Rationale.

### The rebase

`dev` had moved: a sibling ticket landed the optimiser half of the same mask. Three things follow, and all three are in this commit.

  - `expand_investable_weights` is that ticket's, so the two methods this one had added to `src/13_Prior/01_Base_Prior.jl` are deleted. They were duplicates, and a duplicate method refuses to precompile.
  - `NaiveOptimisationResult`, `StackingResult` and `SubsetResamplingResult` gained an `imsk` field there, so each gains a `result_investable_mask` method here.
  - The `investable_reduction` entry of `docs/src/api/20_Optimisation/01_Base_Optimisation.md` is qualified by its signature.

### What this leaves

The second half of #673 and the remaining decisions of #674 are untouched by this ticket.

**Comment by dcelisgarza** [comment]
Resolved by commit e968f2624f on dev. The resolution is the comment above.
